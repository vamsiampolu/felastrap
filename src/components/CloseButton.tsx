import type { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";
import closeIcon from "/close-button.svg";
import { StyleProps, Theme } from "./common";

const close = () => {
  const borderRadius = {
    borderTopLeftRadius: "0.375rem",
    borderTopRightRadius: "0.375rem",
    borderBottomRightRadius: "0.375rem",
    borderBottomLeftRadius: "0.375rem",
  };

  return {
    ...borderRadius,
    borderColor: "rgb(0,0,0)",
    backgroundColor: "transparent",
    backgroundImage: `url(${closeIcon})`,
    backgroundPositionX: "center",
    backgroundOrigin: "padding-box",
    backgroundClip: "border-box",
    backgroundPositionY: "1em",
    backgroundRepeat: "no-repeat",
    backgroundSize: "auto",
    boxSizing: "content-box",
    border: 0,
    color: "#000",
    height: "1em",
    opacity: 0.5,
    paddingTop: "0.25em",
    paddingBottom: "0.25em",
    paddingLeft: "0.25em",
    paddingRight: "0.25em",
    width: "1em",
    ":focus": {
      boxShadow: "0 0 0 0.25rem rgba(13, 110, 253, 0.25)",
      opacity: 1,
      outline: 0,
    },
    ":disabled": {
      opacity: 0.25,
      pointerEvents: "none",
      userSelect: "none",
    },
  };
};

const alertRule = ({ alert }: StyleProps<CloseButtonProps>) => {
  return alert
    ? {
        paddingTop: "1.25rem",
        paddingBottom: "1.25rem",
        paddingLeft: "1rem",
        paddingRight: "1rem",
        position: "absolute",
        right: 0,
        top: 0,
        zIndex: 2,
      }
    : {};
};

const modalRule = ({ modal }: StyleProps<CloseButtonProps>) =>
  modal
    ? {
        padding: "0.5rem 0.5rem",
        margin: "-0.5rem -0.5rem -0.5rem auto",
      }
    : {};

export type CloseButtonProps = ComponentPropsWithoutRef<"button"> &
  Partial<{ alert?: boolean; modal?: boolean }>;

export const CloseButton = ({
  alert = false,
  modal = false,
  className = "",
  ...other
}: CloseButtonProps) => {
  const { css } = useFela<Theme, CloseButtonProps>({ alert, modal });
  const cx = `${css([
    close,
    alertRule,
    modalRule,
  ])} close-button ${className}`.trim();
  return (
    <button
      data-testid="close-button"
      className={cx}
      aria-label="Close"
      {...other}
    />
  );
};
