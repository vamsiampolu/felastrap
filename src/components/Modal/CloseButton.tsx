import { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";
import close from "/close-button.svg";

const rule = () => ({
  marginTop: "-2px",
  appearance: "none",
  textAlign: "end",
  cursor: "pointer",
  border: 0,
  position: "absolute",
  right: "2rem",
  top: "1rem",
  padding: "1rem",
  backgroundColor: "transparent",
  backgroundImage: `url(${close})`,
  backgroundRepeat: "no-repeat",
  backgroundSize: "16px",
  backgroundPosition: "center",
});

export type CloseButtonProps = Omit<
  ComponentPropsWithoutRef<"button">,
  "type" | "aria-label"
>;

export const CloseButton = ({
  onClick,
  className = "",
  ...props
}: CloseButtonProps) => {
  const { css } = useFela();
  const cx = `${css(rule)} bs-close ${className}`.trim();
  return (
    <button
      data-testid="modal-close-button"
      className={cx}
      onClick={onClick}
      type="button"
      aria-label="Close"
      {...props}
    ></button>
  );
};
