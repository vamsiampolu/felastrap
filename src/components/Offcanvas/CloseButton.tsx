import { useFela } from "react-fela";
import closeIcon from "/close-button.svg";
import { ComponentPropsWithoutRef } from "react";

const borderRadius = {
  borderTopLeftRadius: "0.375rem",
  borderTopRightRadius: "0.375rem",
  borderBottomRightRadius: "0.375rem",
  borderBottomLeftRadius: "0.375rem",
};

const rule = () => ({
  ...borderRadius,
  borderColor: "rgb(0,0,0)",
  backgroundColor: "transparent",
  backgroundImage: `url(${closeIcon})`,
  backgroundPositionX: "center",
  backgroundOrigin: "padding-box",
  backgroundClip: "border-box",
  backgroundPositionY: "1em",
  backgroundRepeat: "no-repeat",
  backgroundSize: "auto",
  boxSizing: "content-box",
  border: 0,
  color: "#000",
  height: "1em",
  opacity: 0.5,
  paddingTop: "10px",
  paddingBottom: "10px",
  paddingLeft: "0.25em",
  paddingRight: "0.25em",
  width: "1em",
  ":focus": {
    boxShadow: "0 0 0 0.25rem rgba(13, 110, 253, 0.25)",
    opacity: 1,
    outline: 0,
  },
  ":disabled": {
    opacity: 0.25,
    pointerEvents: "none",
    userSelect: "none",
  },
});

const offcanvasRule = () => ({
  padding: "calc(1rem * 0.5) calc(1rem * 0.5)",
  marginTop: "calc(-0.5 * 1rem)",
  marginRight: "calc(-0.5 * 1rem)",
  marginBottom: "calc(-0.5 * 1rem)",
});

export type CloseButtonProps = Omit<
  ComponentPropsWithoutRef<"button">,
  "aria-label" | "type" | "children"
>;

export const CloseButton = ({ className = "", ...props }: CloseButtonProps) => {
  const { css } = useFela();
  const cx = `${css([
    rule,
    offcanvasRule,
  ])} offcanvas-btn-close ${className}`.trim();
  return (
    <button
      data-testid="offcanvas-close-button"
      type="button"
      className={cx}
      aria-label="Close"
      {...props}
    />
  );
};
