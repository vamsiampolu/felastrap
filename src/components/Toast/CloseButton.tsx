import { useFela } from "react-fela";
import closeIcon from "/close-button.svg";
import { ComponentPropsWithoutRef } from "react";

const borderRadius = {
  borderTopLeftRadius: "0.375rem",
  borderTopRightRadius: "0.375rem",
  borderBottomRightRadius: "0.375rem",
  borderBottomLeftRadius: "0.375rem",
};

const rule = () => {
  return {
    ...borderRadius,
    marginRight: "calc(-0.5 * 0.75rem)",
    marginLeft: "0.75rem",
    paddingTop: "10px !important",
    paddingBottom: "10px !important",
    borderColor: "rgb(0,0,0)",
    backgroundColor: "transparent",
    backgroundImage: `url(${closeIcon})`,
    backgroundPositionX: "center",
    backgroundOrigin: "padding-box",
    backgroundClip: "border-box",
    backgroundPositionY: "1em",
    backgroundRepeat: "no-repeat",
    backgroundSize: "auto",
    boxSizing: "content-box",
    border: 0,
    color: "#000",
    height: "1em",
    opacity: 0.5,
    width: "1em",
    ":focus": {
      boxShadow: "0 0 0 0.25rem rgba(13, 110, 253, 0.25)",
      opacity: 1,
      outline: 0,
    },
    ":disabled": {
      opacity: 0.25,
      pointerEvents: "none",
      userSelect: "none",
    },
  };
};

export type CloseButtonProps = Omit<
  ComponentPropsWithoutRef<"button">,
  "aria-label" | "type" | "children"
>;
export const CloseButton = ({ className = "", ...props }: CloseButtonProps) => {
  const { css } = useFela();
  const cx = `${css(rule)} toast-btn-close ${className}`;
  return (
    <button
      data-testid="toast-close-button"
      type="button"
      className={cx}
      aria-label="Close"
      {...props}
    />
  );
};
