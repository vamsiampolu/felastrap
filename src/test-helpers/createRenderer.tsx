import expandShorthand from "fela-plugin-expand-shorthand";
import sortClassNames from "fela-sort-classnames";
import sortMediaQueryMobileFirst from "fela-sort-media-query-mobile-first";
import prefixer from "fela-plugin-prefixer";
import placeholderPrefixer from "fela-plugin-placeholder-prefixer";
import { createRenderer as createFelaRenderer, type IRenderer } from "fela";
import { lightTheme } from "../reboot/lightTheme";

export const createRenderer = () => {
  const autoMergeLonghands = true;
  const enhancers = [sortClassNames(), sortMediaQueryMobileFirst()];
  const plugins = [
    prefixer(),
    placeholderPrefixer(),
    expandShorthand(autoMergeLonghands),
  ];

  return createFelaRenderer({
    plugins,
    enhancers,
  });
};

export const cssReboot = (renderer: IRenderer) => {
  renderer.renderStatic(lightTheme);
};
