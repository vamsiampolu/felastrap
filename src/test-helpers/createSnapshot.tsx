import { IRenderer } from "fela";
import { renderToString } from "fela-tools";
import { ReactElement, ReactNode } from "react";

import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import {
  type UserEvent,
  type UserEventApi,
} from "@testing-library/user-event/dist/types/setup/setup";
import * as prettier from "prettier";
import { renderToStaticMarkup } from "react-dom/server";
import { Wrapper } from "./Wrapper";

async function formatCSS(css: string) {
  return prettier.format(css, {
    parser: "css",
    useTabs: false,
    tabWidth: 2,
  });
}

function formatHTML(str: string) {
  return prettier.format(str, { parser: "html" });
}

function renderToMarkup(component: ReactElement): string {
  return renderToStaticMarkup(component);
}

export const createSnapshot = async ({
  component,
  renderer,
  theme = {},
}: {
  component: ReactNode;
  renderer: IRenderer;
  theme?: object;
}) => {
  // reset renderer to have a clean setup
  renderer.clear();

  const element = (
    <Wrapper theme={theme} renderer={renderer}>
      {component}
    </Wrapper>
  );

  const view = renderToMarkup(element);
  const str = renderToString(renderer);

  const css = await formatCSS(str);
  const html = await formatHTML(view);

  return `${css}\n\n${html}`;
};

export type EventName = keyof UserEventApi | "focus" | "focus-visible";
export type CreateUserEventType = {
  eventName: EventName;
  component: ReactElement;
  renderer: IRenderer;
  testId: string;
  theme?: object;
  donotWrap?: boolean;
};

export type SetupEvent = {
  user: UserEvent;
  eventName: EventName;
  element: HTMLElement;
};
const setupEvent = async ({ user, eventName, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.hover(element);
      break;
    }

    case "focus": {
      element.focus();
      break;
    }

    case "focus-visible": {
      element.contentEditable = "true";
      element.focus();
      element.contentEditable = "false";
      break;
    }
  }
};

const clearEvent = async ({ eventName, user, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.unhover(element);
      break;
    }

    case "focus":
    case "focus-visible": {
      element.blur();
      break;
    }
  }
};

export const createUserEventSnapshot = async ({
  eventName,
  component,
  testId,
  renderer,
  theme = {},
  donotWrap = false,
}: CreateUserEventType) => {
  renderer.clear();

  const user = userEvent.setup();
  let wrappedComponent = component;
  if (!donotWrap) {
    wrappedComponent = (
      <Wrapper theme={theme} renderer={renderer}>
        {component}
      </Wrapper>
    );
  }

  render(wrappedComponent);

  const element = await screen.findByTestId(testId);

  setupEvent({ element, eventName, user });
  const view = await renderToMarkup(wrappedComponent);
  const utils = renderToString(renderer);
  const css = await formatCSS(utils);
  const html = await formatHTML(view);

  clearEvent({ eventName, element, user });

  return `${css}\n\n${html}`;
};
