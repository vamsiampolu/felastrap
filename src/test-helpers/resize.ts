export const resizeWidth = (width: number) => {
  window.innerWidth = width;
  dispatchEvent(new Event("resize"));
};

export const resizeHeight = (height: number) => {
  window.innerHeight = height;
  dispatchEvent(new Event("resize"));
};
