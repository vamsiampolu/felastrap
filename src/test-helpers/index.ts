import { Wrapper } from "./Wrapper";
import { createRenderer, cssReboot } from "./createRenderer";
import { createSnapshot, createUserEventSnapshot } from "./createSnapshot";
import { resizeHeight, resizeWidth } from "./resize";

export {
  createRenderer,
  cssReboot,
  createSnapshot,
  createUserEventSnapshot,
  resizeWidth,
  resizeHeight,
  Wrapper,
};
