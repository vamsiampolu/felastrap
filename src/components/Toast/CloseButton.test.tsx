import { IRenderer } from "fela";
import { CloseButton } from "./CloseButton";
import { createRenderer, createSnapshot, cssReboot } from "../../test-helpers";

describe("components/Toast/CloseButton", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("render a close button", async () => {
    const component = <CloseButton />;
    const snapshot = await createSnapshot({ renderer, component });
    expect(snapshot).toMatchSnapshot();
  });

  it("render a close button with a className", async () => {
    const component = <CloseButton className="toast-close" />;
    const snapshot = await createSnapshot({ renderer, component });
    expect(snapshot).toMatchSnapshot();
  });
});
