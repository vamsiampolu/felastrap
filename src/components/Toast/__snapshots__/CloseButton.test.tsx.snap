// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Toast/CloseButton render a close button 1`] = `
".a {
  border-top-left-radius: 0.375rem;
}
.b {
  border-top-right-radius: 0.375rem;
}
.c {
  border-bottom-right-radius: 0.375rem;
}
.d {
  border-bottom-left-radius: 0.375rem;
}
.e {
  margin-right: calc(-0.5 * 0.75rem);
}
.f {
  margin-left: 0.75rem;
}
.g {
  padding-top: 10px !important;
}
.h {
  padding-bottom: 10px !important;
}
.i {
  -webkit-background-color: transparent;
  background-color: transparent;
}
.j {
  -webkit-background-image: url(test-file-stub);
  background-image: url(test-file-stub);
}
.k {
  -webkit-background-position-x: center;
  background-position-x: center;
}
.l {
  background-origin: padding-box;
}
.m {
  background-clip: border-box;
}
.n {
  -webkit-background-position-y: 1em;
  background-position-y: 1em;
}
.o {
  background-repeat: no-repeat;
}
.p {
  background-size: auto;
}
.q {
  box-sizing: content-box;
}
.r {
  color: #000;
}
.s {
  height: 1em;
}
.t {
  opacity: 0.5;
}
.u {
  width: 1em;
}
.y:disabled {
  opacity: 0.25;
}
.z:disabled {
  pointer-events: none;
}
.ab:disabled {
  user-select: none;
}
.ac {
  border-top-width: 0;
}
.ae {
  border-right-width: 0;
}
.af {
  border-bottom-width: 0;
}
.ag {
  border-left-width: 0;
}
.ah {
  border-top-color: rgb(0, 0, 0);
}
.ai {
  border-right-color: rgb(0, 0, 0);
}
.aj {
  border-bottom-color: rgb(0, 0, 0);
}
.ak {
  border-left-color: rgb(0, 0, 0);
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  opacity: 1;
}
.x:focus {
  outline-width: 0;
}


<button
  data-testid="toast-close-button"
  type="button"
  class="a ab ac ae af ag ah ai aj ak b c d e f g h i j k l m n o p q r s t u v w x y z toast-btn-close"
  aria-label="Close"
></button
>;
"
`;

exports[`components/Toast/CloseButton render a close button with a className 1`] = `
".a {
  border-top-left-radius: 0.375rem;
}
.b {
  border-top-right-radius: 0.375rem;
}
.c {
  border-bottom-right-radius: 0.375rem;
}
.d {
  border-bottom-left-radius: 0.375rem;
}
.e {
  margin-right: calc(-0.5 * 0.75rem);
}
.f {
  margin-left: 0.75rem;
}
.g {
  padding-top: 10px !important;
}
.h {
  padding-bottom: 10px !important;
}
.i {
  -webkit-background-color: transparent;
  background-color: transparent;
}
.j {
  -webkit-background-image: url(test-file-stub);
  background-image: url(test-file-stub);
}
.k {
  -webkit-background-position-x: center;
  background-position-x: center;
}
.l {
  background-origin: padding-box;
}
.m {
  background-clip: border-box;
}
.n {
  -webkit-background-position-y: 1em;
  background-position-y: 1em;
}
.o {
  background-repeat: no-repeat;
}
.p {
  background-size: auto;
}
.q {
  box-sizing: content-box;
}
.r {
  color: #000;
}
.s {
  height: 1em;
}
.t {
  opacity: 0.5;
}
.u {
  width: 1em;
}
.y:disabled {
  opacity: 0.25;
}
.z:disabled {
  pointer-events: none;
}
.ab:disabled {
  user-select: none;
}
.ac {
  border-top-width: 0;
}
.ae {
  border-right-width: 0;
}
.af {
  border-bottom-width: 0;
}
.ag {
  border-left-width: 0;
}
.ah {
  border-top-color: rgb(0, 0, 0);
}
.ai {
  border-right-color: rgb(0, 0, 0);
}
.aj {
  border-bottom-color: rgb(0, 0, 0);
}
.ak {
  border-left-color: rgb(0, 0, 0);
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  opacity: 1;
}
.x:focus {
  outline-width: 0;
}


<button
  data-testid="toast-close-button"
  type="button"
  class="a ab ac ae af ag ah ai aj ak b c d e f g h i j k l m n o p q r s t u v w x y z toast-btn-close toast-close"
  aria-label="Close"
></button
>;
"
`;
